### PR TITLE
[SPARK-14255] [SQL] Streaming Aggregation

### PR DESCRIPTION
**Pull request by marmbrus**
This PR adds the ability to perform aggregations inside of a `ContinuousQuery`.  In order to implement this feature, the planning of aggregation has augmented with a new `StatefulAggregationStrategy`.  Unlike batch aggregation, stateful-aggregation uses the `StateStore` (introduced in #11645) to persist the results of partial aggregation across different invocations.  The resulting physical plan performs the aggregation using the following progression:
- Partial Aggregation
- Shuffle
- Partial Merge (now there is at most 1 tuple per group)
- StateStoreRestore (now there is 1 tuple from this batch + optionally one from the previous)
- Partial Merge (now there is at most 1 tuple per group)
- StateStoreSave (saves the tuple for the next batch)
- Complete (output the current result of the aggregation)

The following refactoring was also performed to allow us to plug into existing code:
- The get/put implementation is taken from #12013
- The logic for breaking down and de-duping the physical execution of aggregation has been move into a new pattern `PhysicalAggregation`
- The `AttributeReference` used to identify the result of an `AggregateFunction` as been moved into the `AggregateExpression` container.  This change moves the reference into the same object as the other intermediate references used in aggregation and eliminates the need to pass around a `Map[(AggregateFunction, Boolean), Attribute]`.  Further clean up (using a different aggregation container for logical/physical plans) is deferred to a followup.
- Some planning logic is moved from the `SessionState` into the `QueryExecution` to make it easier to override in the streaming case.
- The ability to write a `StreamTest` that checks only the output of the last batch has been added to simulate the future addition of output modes.

### CODE REVIEW

**Comment by yhuai** [comment]
LGTM

**Comment by marmbrus** [comment]
Test this please

**Comment by marmbrus** [comment]
test this please

**Comment by marmbrus** [comment]
@tdas the maintenance test failed twice on this PR, but not the third time.

**Comment by yhuai** [comment]
Aggregation part looks good to me.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/package.scala [hunk 1]
```diff
@@ -28,37 +28,36 @@ package object state {
   implicit class StateStoreOps[T: ClassTag](dataRDD: RDD[T]) {
 
     /** Map each partition of a RDD along with data in a [[StateStore]]. */
-    def mapPartitionWithStateStore[U: ClassTag](
-        storeUpdateFunction: (StateStore, Iterator[T]) => Iterator[U],
+    def mapPartitionsWithStateStore[U: ClassTag](
```
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/package.scala]: nit: isnt it technically more correct to say `mapPartitionsWithStateStore*s*`

File: sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala [hunk 2]
```diff
@@ -224,11 +239,8 @@ trait StreamTest extends QueryTest with Timeouts {
          """.stripMargin
 
     def verify(condition: => Boolean, message: String): Unit = {
-      try {
-        Assertions.assert(condition)
-      } catch {
-        case NonFatal(e) =>
-          failTest(message, e)
+      if (!condition) {
```
Comment (marmbrus) [inline sql/core/src/test/scala/org/apache/spark/sql/StreamTest.scala]: All uses of `verify` are just doing equality checks with variables (and thus can't throw), except for those that were modified specifically such that that were going to throw exceptions upon failure.  So I think really the problem is overloading what was a simple assert to be an error handler.

The issue this this construction is it now pollutes the output with the obvious:
```
condition was false
[info]   org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:500)
[info]          org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1555)
[info]          org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:466)
[info]          org.apache.spark.sql.StreamTest$class.verify$1(StreamTest.scala:228)
[info]          org.apache.spark.sql.StreamTest$$anonfun$testStream$1.apply(StreamTest.scala:355)
[info]          org.apache.spark.sql.StreamTest$$anonfun$testStream$1.apply(StreamTest.scala:271)
[info]          scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
[info]          scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
[info]          org.apache.spark.sql.StreamTest$class.testStream(StreamTest.scala:271)
[info]          org.apache.spark.sql.streaming.StreamSuite.testStream(StreamSuite.scala:24)
```

**Comment by marmbrus** [comment]
Thanks, I'm going to merge to master and will address further comments in follow-ups.

**Pull request closed without merging**